File: platform/repository/rest-client/src/main/java/org/tura/platform/repository/client/rest/RestClientRepository.java
/*
 * Tura - Application generation solution
 *
 * Copyright 2008-2023 2182342 Ontario Inc ( arseniy.isakov@turasolutions.com )
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tura.platform.repository.client.rest;

import java.net.URL;
import java.net.URLEncoder;
import java.util.List;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.Invocation;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.tura.platform.datacontrol.commons.IdDeserializationProblemHandler;
import org.tura.platform.datacontrol.commons.ObjectMapperBuilder;
import org.tura.platform.datacontrol.commons.OrderCriteria;
import org.tura.platform.datacontrol.commons.SearchCriteria;
import org.tura.platform.repository.core.Repository;
import org.tura.platform.repository.core.RepositoryException;
import org.tura.platform.repository.core.SearchRequest;
import org.tura.platform.repository.core.SearchResult;
import org.tura.platform.repository.core.UpdateRequest;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.databind.ObjectMapper;

public class RestClientRepository implements Repository {

	private static final long serialVersionUID = -1003783682499426763L;
	private URL base;
	private Client client;
	private  RequestCustomizer customizer;

	public RestClientRepository(URL base, RequestCustomizer customizer) {
		this.base = base;
		this.customizer = customizer;
	}

	public RestClientRepository(URL base) {
		this.base = base;
		this.customizer = null;
	}

	
	@Override
	public <T> T create(Class<T> objectClass) throws RepositoryException {
		try {
			client = ClientBuilder.newClient();
			ObjectMapper mapper = ObjectMapperBuilder.getObjectMapper();

			String context = base.getPath();
			
			 Invocation.Builder builder = client.target(new URL(base, context + "rest/repository/create").toExternalForm())
						.path("{id}").resolveTemplate("id", URLEncoder.encode(objectClass.getName(), "UTF-8"))
						.request(MediaType.APPLICATION_JSON);
			 if ( customizer != null) {
				 builder = customizer.process(builder);
			 }
			
			Response response = builder.get(Response.class);

			if (response.getStatus() == Response.Status.OK.getStatusCode()) {
				String json = response.readEntity(String.class);
				return mapper.readValue(json,objectClass );
			} else {
				throw new RepositoryException(response.readEntity(String.class));
			}
		} catch (Exception e) {
			throw new RepositoryException(e);
		}finally{
			client.close();
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public <T> SearchResult<T> find(List<SearchCriteria> searchCriteria, List<OrderCriteria> orderCriteria, Integer startIndex,
			Integer endIndex, Class<T> objectClass) throws RepositoryException {

		try {
			ObjectMapper mapper = ObjectMapperBuilder.getObjectMapper();
			mapper.enableDefaultTyping(ObjectMapper.DefaultTyping.OBJECT_AND_NON_CONCRETE,
					JsonTypeInfo.As.PROPERTY);
			mapper.addHandler(new IdDeserializationProblemHandler());

			client = ClientBuilder.newClient();

			SearchRequest request = new SearchRequest();
			request.setSearch(searchCriteria);
			request.setOrder(orderCriteria);
			request.setStartIndex(startIndex);
			request.setEndIndex(endIndex);
			request.setObjectClass(objectClass.getName());

			String context = base.getPath();
			 Invocation.Builder builder = client.target(new URL(base, context + "rest/repository/find").toExternalForm())
						.request(MediaType.APPLICATION_JSON);
			 if ( customizer != null) {
				 builder = customizer.process(builder);
			 }
			Response response = builder.post(Entity.entity(mapper.writeValueAsBytes(request), MediaType.APPLICATION_JSON));

			if (response.getStatus() != Response.Status.OK.getStatusCode()) {
				throw new RepositoryException(response.readEntity(String.class));
			}

			String map = response.readEntity(String.class);
			
			
			SearchResult<T> result = mapper.readValue(map.getBytes(), SearchResult.class);
			
			return result;
		} catch (Exception e) {
			throw new RepositoryException(e);
		}finally{
			client.close();
		}
		
	}

	@Override
	public <T>  void  insert(Object obj, Class<T> objectClass) throws RepositoryException {
		throw new UnsupportedOperationException();

	}

	@Override
	public <T>  void remove(Object obj, Class<T> objectClass) throws RepositoryException {
		throw new UnsupportedOperationException();

	}

	@Override
	public List<Object> applyChanges( List<Object> changes) throws RepositoryException {
		try {
			client = ClientBuilder.newClient();

			ObjectMapper mapper = ObjectMapperBuilder.getObjectMapper();
			mapper.enableDefaultTyping(ObjectMapper.DefaultTyping.OBJECT_AND_NON_CONCRETE,
					JsonTypeInfo.As.PROPERTY);
			mapper.addHandler(new IdDeserializationProblemHandler());

			UpdateRequest request = new UpdateRequest();
			request.setChange(changes);

			String context = base.getPath();
			
			Invocation.Builder builder = client.target(new URL(base, context + "rest/repository/applyChanges").toExternalForm())
					.request(MediaType.APPLICATION_JSON);
			 if ( customizer != null) {
				 builder = customizer.process(builder);
			 }
		    Response response = builder.post(Entity.entity( mapper.writeValueAsBytes(request), MediaType.APPLICATION_JSON));

			
			if (response.getStatus() != Response.Status.OK.getStatusCode()) {
				throw new RepositoryException(response.readEntity(String.class));
			}
			
			String map = response.readEntity(String.class);
			UpdateRequest resp = mapper.readValue(map.getBytes(), UpdateRequest.class);
			
			return resp.getChange();
		} catch (Exception e) {
			throw new RepositoryException(e);
		}finally{
			client.close();
		}
	}

	@Override
	public  <T> T find(Object pk, Class<T> objectClass) throws RepositoryException {
		throw new UnsupportedOperationException();
	}

	
	
}

File: platform/platform-ts/repository/repo-client/src/main/ts/RestClientRepository.ts
import {DataProvider, Repository, SearchRequest, SearchResult} from "repo-core/repo-core";
import {OrderCriteria, SearchCriteria} from "dc-commons/dc-commons";
import {deserialize, serialize} from "json-typescript-mapper/index";


export interface Client{
    target (url:string):Client;
    path(path : string):Client;
    request( type:string):Client;
    post(type : any):Response;
    get():Response;
}

interface Response{
    getStatus() :ResponseStatus;
    getBody() : JSON;
}

enum ResponseStatus{
    OK,
    ERROR
}


export class RestClientRepository implements Repository {

    private client : Client;
    private base : string;


    public RestClientRepository(base : string, client : Client){
        this.base = base;
        this.client=client;
    }

    public addProvider(provider: DataProvider, objectClass: string): any{
        throw new Error("Unsupported operation");
    }

    public addCommand(provider: DataProvider, dataClass: string): any{
        throw new Error("Unsupported operation");
    }

    public create(objectClass: string): any{
        var obj : any ;

        eval("obj=new " + objectClass + "()");
        var response : Response = this.client.target(this.base + "rest/repository/create")
            .path(objectClass).request("application/json").get();

        if (response.getStatus() == ResponseStatus.OK ){
            response.getBody()
            return deserialize( obj,response.getBody());
        }else{
            throw new Error("Bad response");
        }


    }

    public find(searchCriteria: java.util.List<SearchCriteria>, orderCriteria: java.util.List<OrderCriteria>, startIndex: number, endIndex: number, objectClass: string): SearchResult{

        var request :SearchRequest = new SearchRequest();
        request.setSearch(searchCriteria);
        request.setOrder(orderCriteria);
        request.setStartIndex(startIndex);
        request.setEndIndex(endIndex);
        request.setObjectClass(objectClass);

        for( let i= 0; i <  searchCriteria.size(); i++ ) {
            let sc = searchCriteria.get(i);
            sc.setClassName( sc.getValue().getClass().getName());
            sc.setValue(sc.getValue().toString());
        }

        var response : Response = this.client.target(this.base+ "rest/repository/find")
            .request("application/json").post(serialize(request));

        if (response.getStatus() == ResponseStatus.OK ){
           var list : java.util.ArrayList<any> = new java.util.ArrayList();
           var json:JSON =  response.getBody();

           for ( let key in json ){
               let serialObj: JSON = json[key];
               let clazz  = json[key+"_type"];
               list.add(deserialize(clazz,serialObj));
           }
           var size =  json["size"];

           var result : SearchResult = new SearchResult(list,size);
           return result;

        }else{
            throw new Error("Bad response");
        }
    }

    public insert(obj: any, objectClass: string): any{
        throw new Error("Unsupported operation");
    }

    public remove(obj: any, objectClass: string): any{
        throw new Error("Unsupported operation");
    }

    public applyChanges(changes: java.util.List<any>): any{


        var arrayofChanges  ={};
        var rows = 0;

        for( let i= 0; i <  changes.size(); i++ ) {

            let sc = changes.get(i);
            let changeRow = {};

            changeRow[0]=i+"_type";
            changeRow[1]=sc.getClass();
            arrayofChanges[rows]=changeRow;
            rows++;

            changeRow = {};

            changeRow[0]=i;
            changeRow[1]=sc;
            arrayofChanges[rows]=changeRow;
            rows++;
        }

        var response : Response = this.client.target(this.base + "rest/repository/applyChanges")
            .request("application/json").post(serialize(arrayofChanges));

        if (response.getStatus() != ResponseStatus.OK ){
            throw new Error("Bad response");
        }
    }
}